### DISCUSSION TITLE
YAML / JSON data access in custom theme' templates

### DISCUSSION THREAD

**Discussion by redtide** [mkdocs/mkdocs#3239] [Q&A] [answered]
IIUC macros plugin provides access to YAML / JSON data as variables to `docs_dir` (documentation), but these are not accessible in `custom_dir` (theming).
Which possibilities are there to achieve such requirement? Given that I don't want to incorporate data in front matters nor inside the config file but as external data file.

**Comment by pawamoy** [comment, marked as answer]
You can add global variables (attributes, functions, whatever) and filters to the Jinja environment with the [`on_env` hook](https://www.mkdocs.org/dev-guide/plugins/#on_env). Something like this:

```yaml
# mkdocs.yml
hooks:
- my_hook.py
```

```python
# my_hook.py
import json

def load_data(path):
    with open(path) as file:
        return json.load(file)

def on_env(env, **kwargs):
    env.globals["load_data"] = load_data
    # or add a filter with env.filters["filter_name"] = filter_func
```

With this I believe you'll be able to use `load_data` in custom theme templates:

```django
{{ load_data("path/to/data.json") }}
```

...and use filters, loops, and all the other Jinja expressions with these data.

**Reply by redtide** [reply]
Thank you for the quick and detailed answer! I'll try this ASAP.

**Comment by redtide** [comment]
in summary (for Jekyll users that would like to try/switch to MKDocs):

For the layout Jekyll has `_layout` and `_includes` dirs, MKDocs has [theme: custom_dir](https://www.mkdocs.org/user-guide/customizing-your-theme/#using-the-theme-custom_dir).

While Jekyll let to access data files anywhere, MKDocs has a separation for this, so it's possible to `include` or `import` files using the [macros plugin](https://mkdocs-macros-plugin.readthedocs.io/en/latest/) for documentation, and `include` files for the theme only in the `custom_dir`.

With the solution above it's possible also to load data either from the same data directory used with the macros plugin or anywhere else on the `custom_dir`. 

is this correct?
